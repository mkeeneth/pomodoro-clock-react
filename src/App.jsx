import React from 'react';
import utils from './utils';
import './App.css';

/* ================================
libs: reactjs, reactjs-dom
================================= */

// ==== functional components ====
function Footer() {
  return (
    <div className="row">
      <div className="col-sm created-by">
        Created by
        <a href="https://github.com/mkeeneth">Matt Keeneth</a>
      </div>
    </div>
  );
}

function Controls(props) {
  return (
    <div className="row">
      <div className="col-sm control-buttons">
        <button className="btn btn-primary" type="button" id="start_stop" onClick={props.start}>
          Start / Pause
        </button>
        <button className="btn btn-danger" type="button" id="reset" onClick={props.reset}>
          Reset
        </button>
      </div>
    </div>
  );
}

function Session(props) {
  return (
    <div className="row">
      <div className="col-sm">
        <h2 id="timer-label">{props.displayTitle}</h2>
        <div id="time-left" className="session-time">
          {props.time}
        </div>
      </div>
    </div>
  );
}

function BreakControl(props) {
  return (
    <div className="col-sm increment-buttons" id="break-label">
      <h3>Break Length</h3>
      <button
        className="btn btn-primary btn-sm"
        type="button"
        id="break-decrement"
        onClick={props.decrementBreakTime}
      >
        (-)
      </button>
      <span id="break-length">{props.breakTimeLength}</span>
      <button
        className="btn btn-primary btn-sm"
        type="button"
        id="break-increment"
        href=""
        onClick={props.incrementBreakTime}
      >
        (+)
      </button>
    </div>
  );
}

function SessionControl(props) {
  return (
    <div className="col-sm increment-buttons" id="session-label">
      <h3>Session Length</h3>
      <button
        className="btn btn-primary btn-sm"
        type="button"
        id="session-decrement"
        href=""
        onClick={props.decrementSessionTime}
      >
        (-)
      </button>
      <span id="session-length">{props.sessionTimeLength}</span>
      <button
        className="btn btn-primary btn-sm"
        type="button"
        id="session-increment"
        href=""
        onClick={props.incrementSessionTime}
      >
        (+)
      </button>
    </div>
  );
}

// ==== 'smart' components ====
class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      breakTimeLength: 5,
      sessionTimeLength: 25,
      time: 25 * 60, // seconds
      running: false,
      paused: false,
      displayTitle: 'Session',
    };

    // set this constructor scope
    this.decrementBreakTime = this.decrementBreakTime.bind(this);
    this.decrementSessionTime = this.decrementSessionTime.bind(this);
    this.incrementBreakTime = this.incrementBreakTime.bind(this);
    this.incrementSessionTime = this.incrementSessionTime.bind(this);
    this.start = this.start.bind(this);
    this.reset = this.reset.bind(this);
  }

  // Timer stop/pause & reset
  start() {
    if (!this.state.running) {
      // start or resume we arent running
      let currentTime;
      if (this.state.paused) {
        currentTime = this.state.time;
      } else if (this.state.displayTitle === 'Break') {
        currentTime = this.state.breakTimeLength * 60;
      } else {
        currentTime = this.state.sessionTimeLength * 60;
      }
      this.setState((prevState, props) => ({
        running: true,
        time: currentTime,
      }));

      this.decrementTime = this.decrementTime.bind(this);
      this.timer = setInterval(this.decrementTime, 1000);
    } else {
      // pause
      clearInterval(this.timer);
      this.setState({ running: false, paused: true });
    }
  }

  reset() {
    clearInterval(this.timer);

    this.beep.pause();
    this.beep.currentTime = 0;

    this.setState({
      breakTimeLength: 5,
      sessionTimeLength: 25,
      running: false,
      time: 25 * 60,
      displayTitle: 'Session',
    });
  }

  decrementTime() {
    if (this.state.time > 0) {
      this.setState((prevState, props) => ({
        time: prevState.time - 1,
      }));
    } else {
      // play sound! break time?
      this.beep.play();

      clearInterval(this.timer);

      this.setState((prevState, props) => ({
        time: prevState.breakTimeLength,
        running: false,
        paused: false,
        displayTitle: prevState.displayTitle === 'Session' ? 'Break' : 'Session',
      }));
      this.start();
    }
  }

  // Timer Controls +/-
  decrementBreakTime(e) {
    e.preventDefault();
    if (this.state.breakTimeLength > 1 && !this.state.running) {
      this.setState((prevState, props) => ({
        breakTimeLength: prevState.breakTimeLength - 1,
      }));
    }
  }

  decrementSessionTime(e) {
    e.preventDefault();
    if (this.state.sessionTimeLength > 1 && !this.state.running) {
      this.setState((prevState, props) => ({
        sessionTimeLength: prevState.sessionTimeLength - 1,
        time: (prevState.sessionTimeLength - 1) * 60,
      }));
    }
  }

  incrementBreakTime(e) {
    e.preventDefault();
    if (this.state.breakTimeLength < 60 && !this.state.running) {
      this.setState((prevState, props) => ({
        breakTimeLength: prevState.breakTimeLength + 1,
      }));
    }
  }

  incrementSessionTime(e) {
    e.preventDefault();
    if (this.state.sessionTimeLength < 60 && !this.state.running) {
      this.setState((prevState, props) => ({
        sessionTimeLength: prevState.sessionTimeLength + 1,
        time: (prevState.sessionTimeLength + 1) * 60,
      }));
    }
  }

  render() {
    return (
      <div className="jumbotron text-center">
        <h1>Pomodoro Clock</h1>
        <div className="row">
          <BreakControl
            breakTimeLength={this.state.breakTimeLength}
            decrementBreakTime={this.decrementBreakTime}
            incrementBreakTime={this.incrementBreakTime}
          />
          <SessionControl
            sessionTimeLength={this.state.sessionTimeLength}
            decrementSessionTime={this.decrementSessionTime}
            incrementSessionTime={this.incrementSessionTime}
          />
        </div>
        <Session
          time={utils.timeFormatter(this.state.time)}
          displayTitle={this.state.displayTitle}
        />
        <Controls start={this.start} reset={this.reset} />
        <Footer />
        <audio
          id="beep"
          src="./Morning_Circus.mp3"
          preload="auto"
          ref={(audio) => {
            this.beep = audio;
          }}
        />
      </div>
    );
  }
}

// use in codepen
// ReactDOM.render(<App />, document.getElementById('root'));

// dont use in codepen
export default App;
