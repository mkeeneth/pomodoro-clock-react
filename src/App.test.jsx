import React from 'react';
import ReactDOM from 'react-dom';
import { configure, mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Adapter from 'enzyme-adapter-react-16';
import App from './App';
import {
  Footer, Controls, Session, BreakControl, SessionControl,
} from './App';

configure({ adapter: new Adapter() });

jest.useFakeTimers();

// ==== deep rendering tests ====
test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});

test('contains element with id "start_stop"', () => {
  const component = mount(<App />);
  expect(component.find('#start_stop')).toHaveLength(1);
});

test('#start_stop button is clickable and timer counts down', () => {
  const component = mount(<App />);
  const startTime = component.state().time;
  component.find('#start_stop').simulate('click');
  jest.runOnlyPendingTimers(); // fast forward
  expect(component.state().time).toBeLessThan(startTime); // make sure the time counting down
});

test('#reset button is clickable and resets state', () => {
  const defaultState = {
    breakTimeLength: 5,
    sessionTimeLength: 25,
    time: 25 * 60, // seconds
    running: false,
    paused: false,
    displayTitle: 'Session',
  };
  const component = mount(<App />);
  component.find('#break-increment').simulate('click'); // change values
  component.find('#session-increment').simulate('click'); // change values
  component.find('#start_stop').simulate('click'); // start
  component.find('#reset').simulate('click'); // make sure it reset
  expect(component.state()).toEqual(defaultState);
});

test('#break-increment button is clickable', () => {
  const component = mount(<App />);
  const currentBreakTimeLength = component.state().breakTimeLength;
  component.find('#break-increment').simulate('click');
  expect(component.state().breakTimeLength).toBeGreaterThan(currentBreakTimeLength);
});

test('#break-decrement button is clickable', () => {
  const component = mount(<App />);
  const currentBreakTimeLength = component.state().breakTimeLength;
  component.find('#break-decrement').simulate('click');
  expect(component.state().breakTimeLength).toBeLessThan(currentBreakTimeLength);
});

test('#session-increment button is clickable', () => {
  const component = mount(<App />);
  const currentSessionTimeLength = component.state().sessionTimeLength;
  component.find('#session-increment').simulate('click');
  expect(component.state().sessionTimeLength).toBeGreaterThan(currentSessionTimeLength);
});

test('#session-decrement button is clickable', () => {
  const component = mount(<App />);
  const currentSessionTimeLength = component.state().sessionTimeLength;
  component.find('#session-decrement').simulate('click');
  expect(component.state().sessionTimeLength).toBeLessThan(currentSessionTimeLength);
});

// ==== snapshot tests ===
test('footer components renders properly', () => {
  const component = renderer.create(<Footer />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('Controls components renders properly', () => {
  const component = renderer.create(<Controls />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('Session components renders properly', () => {
  const component = renderer.create(<Session />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('BreakControl components renders properly', () => {
  const component = renderer.create(<BreakControl />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('SessionControl components renders properly', () => {
  const component = renderer.create(<SessionControl />);
  expect(component.toJSON()).toMatchSnapshot();
});
