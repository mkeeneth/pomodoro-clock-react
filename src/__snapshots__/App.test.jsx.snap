// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreakControl components renders properly 1`] = `
<div
  className="col-sm increment-buttons"
  id="break-label"
>
  <h3>
    Break Length
  </h3>
  <button
    className="btn btn-primary btn-sm"
    id="break-decrement"
    onClick={undefined}
    type="button"
  >
    (-)
  </button>
  <span
    id="break-length"
  />
  <button
    className="btn btn-primary btn-sm"
    href=""
    id="break-increment"
    onClick={undefined}
    type="button"
  >
    (+)
  </button>
</div>
`;

exports[`Controls components renders properly 1`] = `
<div
  className="row"
>
  <div
    className="col-sm control-buttons"
  >
    <button
      className="btn btn-primary"
      id="start_stop"
      onClick={undefined}
      type="button"
    >
      Start / Pause
    </button>
    <button
      className="btn btn-danger"
      id="reset"
      onClick={undefined}
      type="button"
    >
      Reset
    </button>
  </div>
</div>
`;

exports[`Session components renders properly 1`] = `
<div
  className="row"
>
  <div
    className="col-sm"
  >
    <h2
      id="timer-label"
    />
    <div
      className="session-time"
      id="time-left"
    />
  </div>
</div>
`;

exports[`SessionControl components renders properly 1`] = `
<div
  className="col-sm increment-buttons"
  id="session-label"
>
  <h3>
    Session Length
  </h3>
  <button
    className="btn btn-primary btn-sm"
    href=""
    id="session-decrement"
    onClick={undefined}
    type="button"
  >
    (-)
  </button>
  <span
    id="session-length"
  />
  <button
    className="btn btn-primary btn-sm"
    href=""
    id="session-increment"
    onClick={undefined}
    type="button"
  >
    (+)
  </button>
</div>
`;

exports[`footer components renders properly 1`] = `
<div
  className="row"
>
  <div
    className="col-sm created-by"
  >
    Created by
    <a
      href="https://github.com/mkeeneth"
    >
      Matt Keeneth
    </a>
  </div>
</div>
`;
